### PR TITLE
GRN2-176: Create a role editor that allows admins to specify what permissions each role has

### PR DESCRIPTION
**Pull request by shawn-higgins1**
I've create this large PR as a preliminary pass at a roles editor. However, there are some other options for roles management:
1) Keep this PR as it is and allow any user with the appropriate permissions to edit roles
2) Only allow super admins to modify roles
3) Remove Rolify and scope the roles to providers but, just hardcore the role permissions
4) Keep Rolify and continue basing the permission off the Role name

It would be great to get some feedback on what option you guys think is the best way forward.

### CODE REVIEW

File: app/assets/javascripts/user_edit.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+$(document).on('turbolinks:load', function(){
```
Comment (farhatahmad) [inline app/assets/javascripts/user_edit.js]: Missing copyright message

File: app/assets/stylesheets/application.scss [hunk 2]
```diff
@@ -27,6 +27,7 @@
  * files in this directory. Styles in this file should be added after the last require_* statement.
  * It is generally better to create a new file per style scope.
  *
+ *= require jquery-ui/sortable
```
Comment (farhatahmad) [inline app/assets/stylesheets/application.scss]: Why not use @import like we do below with tabler and bootstrap?

File: app/controllers/rooms_controller.rb [hunk 3]
```diff
@@ -138,6 +141,27 @@ def destroy
     redirect_to current_user.main_room
   end
 
+  # POST room/join
+  def join_specific_room
+    parsed_url = params[:join_room][:url].split('/')
+
+    room_uid = parsed_url[parsed_url.length - 1]
```
Comment (farhatahmad) [inline app/controllers/rooms_controller.rb]: You can replace this with `URI(params[:join_room][:url]).path.split('/').last`

File: app/controllers/rooms_controller.rb [hunk 4]
```diff
@@ -138,6 +141,27 @@ def destroy
...
+
+    begin
+      @room = Room.find_by(uid: room_uid)
+    rescue ActiveRecord::RecordNotFound
+      flash[:alert] = I18n.t("room.no_room.invalid_room_uid")
+      return redirect_to room_path(current_user.main_room)
```
Comment (farhatahmad) [inline app/controllers/rooms_controller.rb]: You should be able to just `redirect_to current_user.main_room` with the flash in one line

File: app/controllers/rooms_controller.rb [hunk 5]
```diff
@@ -138,6 +141,27 @@ def destroy
...
+    end
+
+    if @room.nil?
+      flash[:alert] = I18n.t("room.no_room.invalid_room_uid")
+      return redirect_to room_path(current_user.main_room)
```
Comment (farhatahmad) [inline app/controllers/rooms_controller.rb]: Same as above

File: app/mailers/user_mailer.rb [hunk 6]
```diff
@@ -35,20 +35,22 @@ def password_reset(user, url, image, color)
...
 
-  def user_promoted(user, url, image, color)
+  def user_promoted(user, role, url, image, color)
     @url = url
     @admin_url = url + "admins"
     @image = image
     @color = color
-    mail to: user.email, subject: t('mailer.user.promoted.subtitle')
+    @role = role
+    mail to: user.email, subject: t('mailer.user.promoted.subtitle', role: role)
```
Comment (farhatahmad) [inline app/mailers/user_mailer.rb]: Since you've set @role, you don't need to pass role as an argument. You should be able to access it directly in the email

File: app/mailers/user_mailer.rb [hunk 7]
```diff
@@ -35,20 +35,22 @@ def password_reset(user, url, image, color)
...
 
-  def user_demoted(user, url, image, color)
+  def user_demoted(user, role, url, image, color)
     @url = url
     @root_url = url
     @image = image
     @color = color
-    mail to: user.email, subject: t('mailer.user.demoted.subtitle')
+    @role = role
+    mail to: user.email, subject: t('mailer.user.demoted.subtitle', role: role)
```
Comment (farhatahmad) [inline app/mailers/user_mailer.rb]: Same as above

File: app/models/user.rb [hunk 8]
```diff
@@ -228,6 +239,53 @@ def self.new_token
...
+  # role functions
+  def highest_priority_role
+    roles.by_priority.first
+  end
+
+  def add_role(role)
+    unless has_role?(role)
+      role_provider = "greenlight"
+
+      role_provider = provider if Rails.configuration.loadbalanced_configuration
```
Comment (farhatahmad) [inline app/models/user.rb]: You can combine this and the line above into 1 conditional statement

File: app/models/user.rb [hunk 9]
```diff
@@ -251,6 +309,12 @@ def initialize_main_room
 
   # Initialize the user to use the default user role
   def assign_default_role
+    role_provider = "greenlight"
+
+    role_provider = provider if Rails.configuration.loadbalanced_configuration
```
Comment (farhatahmad) [inline app/models/user.rb]: Same as previous comment

File: yarn.lock [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (farhatahmad) [inline yarn.lock]: Add this file to the .gitignore

File: app/mailers/user_mailer.rb (hunk 6 above)
Comment (shawn-higgins1) [reply app/mailers/user_mailer.rb]: I'm just passing the role argument to I18n to translate the subject. I user @ROLE in the email view itself.

File: yarn.lock (hunk 10 above)
Comment (shawn-higgins1) [reply yarn.lock]: Shouldn't lock files get committed to the repo?

File: app/mailers/user_mailer.rb (hunk 6 above)
Comment (farhatahmad) [reply app/mailers/user_mailer.rb]: Oh oops. I didn't notice that. You can ignore this comment then

File: yarn.lock (hunk 10 above)
Comment (farhatahmad) [reply yarn.lock]: Yeah you're right. You can ignore this as well

File: app/controllers/admins_controller.rb [hunk 11]
```diff
@@ -36,7 +36,9 @@ def index
     @search = params[:search] || ""
     @order_column = params[:column] && params[:direction] != "none" ? params[:column] : "created_at"
     @order_direction = params[:direction] && params[:direction] != "none" ? params[:direction] : "DESC"
-    @role = params[:role] || ""
+    @role = nil
+
+    @role = Role.find_by(name: params[:role], provider: @user_domain) if params[:role]
```
Comment (farhatahmad) [inline app/controllers/admins_controller.rb]: You can combine this line and the line above it into 1 line with a conditional

File: app/controllers/admins_controller.rb [hunk 12]
```diff
@@ -99,6 +83,7 @@ def unban_user
   # POST /admins/approve/:user_uid
   def approve
     @user.remove_role :pending
+    @user.add_role :user
```
Comment (farhatahmad) [inline app/controllers/admins_controller.rb]: Why is this needed? Isn't the the `user` role already added to the user when they are created?

File: app/controllers/admins_controller.rb [hunk 13]
```diff
@@ -183,6 +168,140 @@ def default_recording_visibility
...
+
+    @selected_role = if params[:selected_role].nil?
+                        @roles.find_by(name: 'user')
+                      else
+                        @roles.find(params[:selected_role])
+                     end
+  end
+
+  # POST /admin/role
+  def new_role
```
Comment (farhatahmad) [inline app/controllers/admins_controller.rb]: Can you add some comments to this function? It'll help explain what is going on exactly

File: app/controllers/admins_controller.rb [hunk 14]
```diff
@@ -183,6 +168,140 @@ def default_recording_visibility
...
+    user_role.priority += 1
+
+    new_role.save!
+    user_role.save!
+
+    redirect_to admin_roles_path(selected_role: new_role.id)
+  end
+
+  # PATCH /admin/roles/order
+  def change_role_order
```
Comment (farhatahmad) [inline app/controllers/admins_controller.rb]: Same as above

File: app/controllers/admins_controller.rb [hunk 15]
```diff
@@ -183,6 +168,140 @@ def default_recording_visibility
...
+      top_priority = new_priority
+      Role.where(id: id).update_all(priority: new_priority)
+    end
+
+    user_role.priority = top_priority + 1
+    user_role.save!
+  end
+
+  # POST /admin/role/:role_id
+  def update_role
```
Comment (farhatahmad) [inline app/controllers/admins_controller.rb]: Same as above

File: app/controllers/admins_controller.rb [hunk 16]
```diff
@@ -183,6 +168,140 @@ def default_recording_visibility
...
+
+    role.update(permission_params)
+
+    role.save!
+
+    redirect_to admin_roles_path(selected_role: role.id)
+  end
+
+  # DELETE admins/role/:role_id
+  def delete_role
```
Comment (farhatahmad) [inline app/controllers/admins_controller.rb]: Same here

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -249,4 +251,53 @@ def passes_invite_reqs
 
     invitation[:present]
   end
+
+  def update_roles
```
Comment (farhatahmad) [inline app/controllers/users_controller.rb]: Can you add some comments here as well please
